PayloadPkg/FirmwareUpdate/FirmwareUpdateHelper.c

```diff
@@ -1151,3 +1160,80 @@ Reboot (
   ResetSystem (ResetType);
   CpuDeadLoop ();
 }
+
+/**
+  Verify uCode internal structure
+
+  @param[in] ImageHdr     Pointer to the fw mgmt capsule image header
+
+  @retval  EFI_SUCCESS    The operation completed successfully.
+  @retval  others         There is error happening.
+**/
+EFI_STATUS
+VerifyUcodeStruct (
+  IN  EFI_FW_MGMT_CAP_IMAGE_HEADER  *ImageHdr
+  )
+{
+  UINTN                 ImageBase;
+  UINTN                 ImageOffset;
+  CPU_MICROCODE_HEADER  *UCodeHdr;
+  UINT8                 *ImageByte;
+
+  // Update is only supported for platforms that slot their uCode
+  if (PcdGet32 (PcdUcodeSlotSize) == 0) {
+    DEBUG((DEBUG_ERROR, "Existing image does not contain uCode slots!!\n"));
+    return EFI_UNSUPPORTED;
+  }
+
+  ImageBase = (UINTN)ImageHdr + sizeof(EFI_FW_MGMT_CAP_IMAGE_HEADER);
+  ImageOffset = 0;
+
+  // Ensure patches in update image start at slot boundaries
+  ImageByte = (UINT8*)(ImageBase + ImageOffset);
+  while (*ImageByte != PAD_BYTE && ImageOffset < ImageHdr->UpdateImageSize) {
+    UCodeHdr = (CPU_MICROCODE_HEADER *)ImageByte;
+    if (UCodeHdr->HeaderVersion != 1) {
+      DEBUG((DEBUG_ERROR, "Existing image slots do not line up with new image slots!!\n"));
+      return EFI_NO_MAPPING;
+    }
+    ImageOffset += PcdGet32(PcdUcodeSlotSize);
+    ImageByte = (UINT8*)(ImageBase + ImageOffset);
+  }
+
+  // Check remaining bytes are unused
+  while (ImageOffset < ImageHdr->UpdateImageSize) {
+    if (*ImageByte != PAD_BYTE) {
+      DEBUG((DEBUG_ERROR, "Existing image slots do not line up with new image slots!!\n"));
+      return EFI_NO_MAPPING;
+    }
+    ++ImageOffset;
+    ImageByte = (UINT8*)(ImageBase + ImageOffset);
+  }
```

Comment:
Please double check if capsule image is bigger (e.g. 3 uCode) than the actual uCode size in flash (e.g. 2 uCode slot available). It would fail to avoid override other flash region.